src/battery-check/battery-check.c

```diff
@@ -0,0 +1,141 @@
+
+        r = loop_write(fd, plymouth_message, c, /* do_poll = */ false);
+        if (r < 0)
+                return (void) log_full_errno(IN_SET(r, -EAGAIN, -ENOENT) || ERRNO_IS_DISCONNECT(r) ?
+LOG_DEBUG : LOG_WARNING, r, "Failed to write to plymouth, ignoring: %m");
+}
+
+static int parse_argv(int argc, char * argv[]) {
+
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",    no_argument, NULL, 'h'         },
+                { "version", no_argument, NULL, ARG_VERSION },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 0);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+
+                switch (c) {
+
+                case 'h':
+                        help();
+                        return 0;
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        if (optind < argc)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "%s takes no argument.",
+                                       program_invocation_short_name);
+        return 1;
+}
+
+static int run(int argc, char *argv[]) {
+        int r;
+
+        log_parse_environment();
+        log_open();
+
+        r = parse_argv(argc, argv);
+        if (r <= 0)
+                return r;
+
+        r = battery_is_discharging_and_low();
+        if (r < 0) {
+                log_warning_errno(r, "Failed to check battery status, ignoring: %m");
+                return 0;
+        }
+        if (r > 0) {
+                _cleanup_close_ int fd = -EBADF;
+                _cleanup_free_ char * message = NULL;
+
+                if (asprintf(&message, "%s Battery level critically low. Please connect your charger or the system powers off in 10 seconds.", special_glyph(SPECIAL_GLYPH_LOW_BATTERY)) < 0)
+                        return log_oom();
+                log_emergency("%s.", message);
+                fd = open_terminal("/dev/console", O_WRONLY|O_NOCTTY|O_CLOEXEC);
+                if (fd < 0)
+                        log_warning_errno(fd, "Failed to open console, ignoring: %m");
+                else
+                        dprintf(fd,  ANSI_HIGHLIGHT_RED "%s" ANSI_NORMAL "\n", message);
+                battery_check_send_plymouth_message(message);
+                sleep(10);
+                r = battery_is_discharging_and_low();
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to check battery status, ignoring: %m");
+                        return 0;
+                }
```

Comment:
> if this returns `r > 0` then print a message that the power was plugged in again, and proceed with `return 0`

This would be if ```r == 0 ```?
To indicate we're either plugged or battery level is back up?

```r > 0 ``` would mean we're still unplugged and discharging?

Reply:
```r < 0 ```: we couldn't read
```r > 0 ```: yes, battery is low and discharging
```r == 0 ``` we're either plugged in or have more than 5% battery charge.

First call to ```battery_is_discharging_and_low()``` returns ```true```, we should print the message, and check again in 10 seconds.

If the user hasn't plugged in, we get ```r > 1``` return it, and force shutdown.

If the user has plugged in, we get ``` r == 0``` print a message that says "A.C power restored, continuing"

If battery level doesn't get read on the second call, we have ```r < 0 ```
whether or not the user has plugged in.

How do we handle this ```r < 0```since we'll be unsure whether or not the user has plugged in?


Sorry it's verbose.

Reply:
yeah i mean r == 0

Reply:
i wouldn't bother too much with the call returning r < 0 the second time. this is not going to happen anyway. hence just do a log_warning() and just assume the second reading is identical to the first, i.e. still battery low. but let's not overthink this, and try to bring this equally loudly to screen, because this is really not an issue that can likely happen. as long as you print a log message *at* all, via log_warning() or so we are good.

Reply:
> i wouldn't bother too much with the call returning r < 0 the second time. this is not going to happen anyway. hence just do a log_warning() and just assume the second reading is identical to the first, i.e. still battery low. but let's not overthink this, and try to bring this equally loudly to screen, because this is really not an issue that can likely happen. as long as you print a log message _at_ all, via log_warning() or so we are good.

Thank you sir.